### ISSUE TITLE
Feature Request: Context and stronger WrapCustomizable

### ISSUE THREAD

**Issue by Schaltfehler** [JohnSundell/Wrap#9]
Hi John!
I love your micro libs Unbox and Wrap but noticed Wrap is missing a few features to make it as awesome as Unbox.

### Context for Wrap:
It would be great to pass a context for Wrap like in Unbox. That way you could define custom context depended keys for properties.  This would come in handy when you have a single object which is use for multiple APIs with different requirements.

### Stronger WrapCustomizable:
With `WrapCustomizable` it is possible to do already a lot but it falls short when i want to flatten out some properties.
For example:
```
struct SpaceShip {
    let weight: Double
    let engine: Engine
}

struct Engine {
    let manufacturer: String
    let fuelConsumption: Float
}
```
I want to turn into
```
{
    "weight": 3999.72,
    "manufacturer": "The Space Company",
    "fuelConsumption": 17.2321
}
```

I could do that with `func wrap() -> AnyObject?` but don't always want to do this for the whole object. Especially when it has a lot of properties and I need to flat-out only a single property. 
Maybe if there would be something `func keyForWrappingPropertyNamed(propertyName: String) -> [String]` where you could return an array of keys...?

**Comment by JohnSundell** [comment]
Hey @Schaltfehler! Happy to hear that you like Unbox and Wrap 😃 Those are both great feature requests, we should make them happen!

The first one is easy; we'll just add a `context` parameter to `Wrap()` just like Unbox has. You're right that it'll make Wrap a lot more powerful.

Second one is a bit trickier, it's almost like you'd want some API that will signal to Wrap that you want to merge a given child model with its parent (like in this case, you're merging the nested `Engine` struct into the wrapped dictionary for `SpaceShip`). How about:
```
protocol WrapCustomizable {
     func shouldMergeWrappedDictionaryForPropertyNamed(propertyName: String) -> Bool
}
```

Let me know what you think and I can either build this, or if you want to contribute :)

**Comment by Schaltfehler** [comment]
Thank you for your reply.

`func shouldMergeWrappedDictionaryForPropertyNamed(propertyName: String) -> Bool` is good. It keeps things simple and keeps the "falvor" of the lib.

Let me recap to make sure i go the idea right and see if there are limitations.
With `func shouldMergeWrappedDictionaryForPropertyNamed(propertyName: String) -> Bool` I only merge the sublayer of PropertyNames into the current layer.

For example:
```
struct SpaceShip {
    let weight: Double
    let engine: Engine
}

struct Engine {
    let manufacturer: Company
    let fuelConsumption: Float
}

struct Company {
    let name: String
    let adrress: String
}
```
If I want to wrap `SpaceShip` and completely flatten its structure, I would make `SpaceShip` and `Engine` implement `shouldMergeWrappedDictionaryForPropertyNamed` and return `true`.
To make this work the deepest layer would need to be ask first, goes up layer by layer and the result would be.
```
{
    "weight": 3999.72,
    "name": "The Space Company",
    "adrress": "Somestreet, 1234, Somewhere",
    "fuelConsumption": 17.2321
}
```
Now we would have everything in the same layer/scope. 
Works for some cases this might be enough but probably not for all of them. 
What is if I want to restructure them again and have a result like this:
```
{
    "specs" : {
        "weight": 3999.72,
        "fuelConsumption": 17.2321
    },
    "info" : {
        "companyName": "The Space Company"
    }
}
```
Please correct me if am wrong but it is probably not possible to do this from this point on?
So how about an other Feature? 
Extending `keyForWrappingPropertyNamed`, by not only letting the return to be a key but a path to the key. Like in `Unbox` if i remember right.
So for example:
```
func keyForWrappingPropertyNamed(propertyName: String) -> String? {
    if propertyName == "weight" {
        return "specs.weight"
    } else if propertyName == "fuelConsumption" {
        return "specs.fuelConsumption"
    } else if propertyName == "name" {
        return "info.companyName"
    } else if propertyName == "adrress" {
        return nil
    }
    return propertyName
}
```
That would be cool and for now it would be fine, I think.

---

Or maybe we could go further. I am not really sure if it would reduce code and add simplicity.
But what if the propertyName of `func keyForWrappingPropertyNamed(propertyName: String) -> String?` would not be just a name but a path as well?
It would be come the center point of translation. What do you think. Would it be too powerfull or somehow against the concept of the lib? It would look like this.
```
func keyForWrappingPropertyNamed(propertyName: String) -> String? {
    if propertyName == "weight" {
        return "specs.weight"
    } else if propertyName == "engine.fuelConsumption" {
        return "specs.fuelConsumption"
    } else if propertyName == "engine.manufacturer.name" {
        return "info.companyName"
    } else if propertyName == "engine.manufacturer.adrress" {
        return nil
    }
    return propertyName
}
```

Let me know what you think :-)

**Comment by kaandedeoglu** [comment]
I would also like to have the option to specify custom paths for the encoding (i.e something like `isKeypath` from `Unbox`).

**Comment by Schaltfehler** [comment]
I played a bit around with different possible implementations of this feature.
One way i tried was to extend the WrapCustomizble extension by adding
`func pathForWrappingPropertyNamed(propertyName: String, root: String) -> String?` which is called for each object in in my model which i want to convert to JSON.
The problem I got with this solution is that each object needs to have knowledge of of the object it is nested in and maybe the object is using in different objects too. Well you can do this by adding a context and letting it know in which object it is nested but maintainability is suffering a lot. For changing or debugging you need to check different objects, keep the whole scheme in your mental memory and so on. 

I might try out a new way to solve this. It might be best to give the Wrapper a Path-Translation-Map right from the  beginning.

For example if we have the following Models:
```
struct SpaceShip {
    let manufacturer: Company
    let engine: Engine
}
struct Engine {
    let manufacturer: Company
    let fuelConsumption: Float
}
struct Company {
    let name: String
    let address: String
}
```
Without custom paths we would get something like this:
```
{
  manufacturer : {
    name: "Wayland",
    address: "London"
  },
  engine : {
    manufacturer : {
      name: "Yutani",
      address: "Tokyo"
    }
    fuelConsumption: 1000
  }
}
```
But lets assume we need to have this scheme:
```
{
  "SpaceShipCompany" : "Weyland",
  "EngineCompany" : "Yutani",
  "Addresses" : [
    "London",
    "Tokyo"
  ]
}
```

Using `func pathForWrappingPropertyNamed(propertyName: String, root: String) -> String?` would require to extend each involved object, know about the overall structure and write code to return the custom path.

The core information needed is just a translation table which maps from the default to the custom path. Something like this:
```
"manufacturer.name" -> "SpaceShipCompany"
"engine.manufacturer.name" -> "EngineCompany"
"manufacturer.address" -> "Addresses"
"engine.manufacturer.address" -> "Addresses"
"engine.fuelConsumption" -> nil
```  

It's probably much easier/practical to define the structure for the json scheme at one place and apply it for Wrapping on the Root object:
```
Wrapper(customPathsMap).wrap(self.mySpaceShip)
```

**Comment by JohnSundell** [comment]
Contexts were implemented in Wrap 2.0
